File: src/amazon_ai/HybridMCTS.java
package amazon_ai;

import static amazon_framework.Operations.*;
import amazon_framework.*;

public class HybridMCTS extends IntelligentAgent {
    int hybrid_hyperparam = 13;
    double max_sim_num;
    private final Thread[] thread_pool = new Thread[32];

    public HybridMCTS(byte[][] board, byte side) {
        super(board, side);
    }

    @Override
    public void calculate() {
        max_sim_num = 35 + 5 * hybrid_hyperparam;
        CoordinateArray valueArray = new CoordinateArray();
        int varr_size = hybrid_hyperparam - 2;
        int maxSimValue = -inf;
        byte[][] bestNode = null;

        // <Expansion>
        // Gets the possible states and calculates available moves for each child state
        MatrixArray childStates = getPossibleStates(gameBoard, side);
        final int size_i = childStates.size();
        for (int i = 0; i < size_i; i++) {
            Coordinate value = new Coordinate();
            value.y += numPossibleMoves(childStates.at(i), side);
            value.y -= numPossibleMoves(childStates.at(i), enemy_side);
            value.x = i;
            valueArray.add(value);
        }
        valueArray.sort_by_y(); // Get largest possible move state list, desending order
        varr_size = min(varr_size, 32);
        // </Expansion>

        // <Selection>
        CoordinateArray selectedNodeCor = valueArray.back(varr_size); // n most valuable moves
        if (selectedNodeCor.isEmpty()) {
            newState = gameBoard;
            return;
        }

        int[] simValue = new int[selectedNodeCor.size()];
        byte[][][] selectedNodeArray = new byte[selectedNodeCor.size()][][];
        for (int i = 0; i < selectedNodeCor.size(); i++) {
            selectedNodeArray[i] = childStates.at(selectedNodeCor.at(i).x);
        }
        // </Selection>

        // <Simulation>
        printf("Max Sim Value = %f\n", max_sim_num);
        printf("HybridMarkov is simulating possible states...\nValue: ");
        for (int i = 0; i < selectedNodeArray.length; i++) {
            final int j = i;
            thread_pool[j] = new Thread(new Runnable() {
                @Override
                public void run() {
                    simValue[j] = randomSimulation(toState(selectedNodeArray[j]));
                }
            });
        }
        for (int i = 0; i < selectedNodeArray.length; i++) {
            thread_pool[i].start();
        }
        for (int i = 0; i < selectedNodeArray.length; i++) {
            try {
                thread_pool[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        // </Simulation>

        for (int i = 0; i < selectedNodeArray.length; i++) {
            if (simValue[i] > maxSimValue) {
                maxSimValue = simValue[i];
                bestNode = matrixCopy(selectedNodeArray[i]);
            }
            if (maxSimValue >= max_sim_num)
                break;
        }
        newState = bestNode;
        hybrid_hyperparam++;
    }

    public int randomSimulation(State s) {
        int value = 0;
        State currentState = s;
        HybridL shadow = new HybridL(s, side);
        HybridS eshadow = new HybridS(s, enemy_side);
        for (int j = 0; j < max_sim_num; j++) {
            currentState = s;
            while (true) {
                currentState.is_opponent = true;
                eshadow.update(currentState);
                eshadow.calculate(true); // Selection and Expansion
                currentState = eshadow.nextNode;
                if (eshadow.getAction().size() <= 0) {
                    value++;
                    eshadow.currentNode.backPropagation(false, 0); //Backpropagation
                    break;
                }
                currentState.is_opponent = false;
                shadow.update(currentState);
                shadow.calculate(false);
                currentState = shadow.nextNode;
                if (shadow.getAction().size() <= 0) {
                    value--;
                    shadow.currentNode.backPropagation(false, 0);
                    break;
                }
            }
        }
        printf("%d ", value);
        return value;
    }
}

File: src/src/cpp/HybridMCTS.cpp
#include "HybridMCTS.h"
#include "HybridSimAgent.h"

void random_simulation(StateNode*, int*, byte, int, int);

void HybridMCTS::calculate()
{
	using std::thread;
	max_sim_num = -170 + 30 * hybrid_hyperparam;
	max_sim_num = min(max_sim_num, 1000);
	int maxSimValue = -INF;
	Matrix bestNode;

	// <Expansion>
	// Gets the possible states and calculates available moves for each child state
	MatrixArray childStatesArr[4];
	OrderedPairList valueArray[4];
	piecePossibleStates(childStatesArr, gameBoard, side);

	for (int i = 0; i < 4; i++)
	{
		const u64 size_j = childStatesArr[i].size();
		valueArray[i].reserve(size_j);
		for (int j = 0; j < size_j; j++) {
			OrderedPair value(0, 0);
			value.y += numPossibleMoves(childStatesArr[i].at(j), side);
			value.y -= numPossibleMoves(childStatesArr[i].at(j), enemy_side);
			value.x = j;
			valueArray[i].push_back(value);
		}
		// largest possible move state list, descending order
		OrderedPairList::sort_by_y(valueArray[i]);
	}
	// </Expansion>

	// <Selection>
	int frontnum = 12;
	for (int i = 0; i < 9; i++)
	{
		if (hybrid_hyperparam - HPSTART > (40 - i * 5))
		{
			frontnum++;
		}
	}

	MatrixArray selectedNodeArray;
	for (int i = 0; i < 4; i++)
	{
		OrderedPairList frontx;
		OrderedPairList::front(frontx, valueArray[i], frontnum);
		for (const OrderedPair& e : frontx)
		{
			selectedNodeArray.push_back(childStatesArr[i].at(e.x));
		}
	}
	const u64 arr_size = selectedNodeArray.size();
	if (selectedNodeArray.empty()) {
		newState = gameBoard;
		return;
	}
	const int sizes = 4 * frontnum * sizeof(int);
	int* simValue = (int*)malloc(sizes * sizeof(int));
	for (int i = 0; i < sizes; i++) {
		simValue[i] = 0;
	}
	// </Selection>


	// <Simulation>
	printf("Max Sim Value = %d\n", max_sim_num);
	printf("Hybrid_MCTS is simulating possible states...\nValue: ");
	std::vector<StateNode*> gc_list;
	for (int i = 0; i < arr_size; i++)
	{
		auto* node = new StateNode(selectedNodeArray[i]);
		thread_pool[i] = thread(random_simulation, node, &(simValue[i]), side, max_sim_num, hybrid_hyperparam - HPSTART);
		gc_list.push_back(node);
	}
	for (int i = 0; i < arr_size; i++)
	{
		thread_pool[i].join();
	}
	// </Simulation>


	for (int i = 0; i < arr_size; i++) {
		if (simValue[i] > maxSimValue) {
			bestNode = selectedNodeArray[i];
			maxSimValue = simValue[i];
		}
	}
	printf("max value = %d\n", maxSimValue);
	newState = bestNode;
	free(simValue);
	for (auto& g : gc_list)
	{
		delete(g);
	}
	hybrid_hyperparam++;
}

void random_simulation(StateNode* s, int* value, byte side, int max_sim_num, int tn)
{
	clock_t start = clock();
	memPool mem_pool;
	const byte enemy_side = (side == WHITE ? BLACK : WHITE);
	HybridSimAgent shadow = HybridSimAgent(s, side, &mem_pool);
	HybridSimAgent eshadow = HybridSimAgent(s, enemy_side, &mem_pool);
	shadow.turn_num = tn;
	eshadow.turn_num = tn;
	for (int j = 0; j < max_sim_num; j++)
	{
		StateNode* currentState = s;
		while (true)
		{
			currentState->opponent = true;
			eshadow.update(currentState);
			// Selection and Expansion
			eshadow.calculate(true);
			currentState = eshadow.nextNode;
			if (eshadow.getAction().empty())
			{
				(*value)++;
				// Backpropagation
				eshadow.currentNode->backPropagation(false, 0);
				break;
			}
			currentState->opponent = false;
			shadow.update(currentState);
			shadow.calculate(false);
			currentState = shadow.nextNode;
			if (shadow.getAction().empty())
			{
				(*value)--;
				shadow.currentNode->backPropagation(false, 0);
				break;
			}
		}
	}
	printf("%d ", *value);
	for (auto* g : shadow.sna_list)
	{
		delete(g);
	}
	for (auto* g : eshadow.sna_list)
	{
		delete(g);
	}
	mem_pool.memfree();
}